theano/tensor/opt.py

```diff
@@ -3401,22 +3546,29 @@ def local_useless_switch(node):
         cond = T.extract_constant(node.inputs[0], elemwise=False)
         if type(cond) is numpy.ndarray and cond.ndim == 0:
             if cond == 0:
-                out = node.inputs[2]
+                correct_out = node.inputs[2]
             else:
-                out = node.inputs[1]
+                correct_out = node.inputs[1]
 
-            if out.ndim != node.outputs[0].ndim:
+            if correct_out.ndim != node.outputs[0].ndim:
                 # TODO: broadcast?
                 return False
-            if out.dtype != node.outputs[0].dtype:
-                out = T.cast(out, node.outputs[0].dtype)
-            if out.type.broadcastable != node.outputs[0].type.broadcastable:
+            if correct_out.dtype != node.outputs[0].dtype:
+                out = T.cast(correct_out, node.outputs[0].dtype)
+            if correct_out.type.broadcastable != node.outputs[0].type.broadcastable:
                 # We need to copy data to the new dimensions during execution
-                out = T.alloc(out, *[node.outputs[0].shape[i] for i
-                                     in xrange(out.ndim)])
+                out = T.alloc(correct_out, *[node.outputs[0].shape[i] for i
+                                     in xrange(correct_out.ndim)])
```

Comment:
If we need to return something else than `correct_out` directly, maybe we should keep the trace of `node.outputs`.
For instance, let's say we have:
```python
x = T.scalar()
y = T.matrix()
z = T.switch(1, x, y)
```
Then `z` will be replace by `alloc(x, ...)`, which could fail because of a memory allocation error. In that case, I guess it would be confusing to report this `alloc` as being created by `T.scalar()`, but it becomes more helpful to trace it back to the `switch` with a large matrix `y`.